api/agent/agent.go

```diff
@@ -764,26 +768,29 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
 	}
 	defer container.Close()
 
-	// XXX(reed): we need to timeout the cookie create / prepare since docker client doesn't have timeout anymore,
-	// and handle cookie close having a timed out context it still needs to delete the thing. fun stuff
 	logger := logrus.WithFields(logrus.Fields{"id": container.id, "app_id": call.AppID, "fn_id": call.FnID, "image": call.Image, "memory": call.Memory, "cpus": call.CPUs, "idle_timeout": call.IdleTimeout})
 	ctx = common.WithLogger(ctx, logger)
 
+	ctx, cancel := context.WithTimeout(common.BackgroundContext(ctx), a.cfg.HotStartTimeout)
+	defer cancel()
+
 	cookie, err := a.driver.CreateCookie(ctx, container)
 	if err != nil {
 		call.slots.queueSlot(&hotSlot{done: make(chan struct{}), fatalErr: err})
 		return
 	}
 
-	defer cookie.Close(ctx)
+	// WARNING: we wait forever.
```

Comment:
retry in docker_client will give up at 10 tries and this should bail at that point, at least, currently I don't think this is true. so long as every call to the docker_client has a timeout in the context (in applicable cases, ie everything but this one?), we could remove the 10 retry cap, as the retry loop respects the context timeout/cancel as it is.